File: archive/c/c/duplicate-character-counter.c
#include <stdio.h>
#include <string.h>
#include<stdlib.h>

// Function to handle errors
int handle_error()
{
    printf("Usage: please provide a string\n");
    exit(0);
}

int main(int argc, char *argv[])
{
    /*
        Condition to check for No String as Input
    */
    if (argc !=2 )
    {
        handle_error();
    }
    /*
        Condition to check for No String as Input
    */
    if(strlen(argv[1]) == 0){
        handle_error();
    }
    int counter[256]={0};
    int len = strlen(argv[1]);
    for(int i=0;i<len;i++){
        counter[argv[1][i]-1]++;
    }
    int flag = 1;
    for(int i=0;i<len;i++){
        char c = argv[1][i];
        if(counter[c-1]>1){
            flag = 0;
            printf("%c: %d\n",c,counter[c-1]);
            counter[c-1]=0;
        }
    }
    if(flag == 1){
        printf("No duplicate characters\n");
    }
    return 0;
}

File: archive/r/rust/duplicate-character-counter.rs
use std::env::args;
use std::process::exit;
use std::collections::{HashMap, HashSet};

fn usage() -> ! {
    println!("Usage: please provide a string");
    exit(0);
}

fn duplicate_character_counter(s: &str) -> HashMap<char, usize> {
    let mut counts: HashMap<char, usize> = HashMap::new();

    // Count number of occurances of each character
    for c in s.chars() {
        *counts.entry(c).or_insert(0) += 1;
    }

    counts
}

fn show_duplicate_character_counts(
    s: &str, counts: &HashMap<char, usize>
) {
    // Show characters that have duplicates and keep track of
    // which duplicate characters are found
    let mut visited: HashSet<char> = HashSet::new();
    for c in s.chars() {
        let count: usize = *counts.get(&c).unwrap();
        if count > 1 && !visited.contains(&c) {
            println!("{c}: {count}");
            visited.insert(c);
        }
    }

    // Indicate if no duplicates found
    if visited.len() < 1 {
        println!("No duplicate characters");
    }
}

fn main() {
    let mut args = args().skip(1);

    // Get 1st command-line argument. Error if empty
    let s: &str = &args
        .next()
        .unwrap_or_else(|| usage());
    if s.len() < 1 {
        usage();
    }

    // Count duplicate characters and show results
    let counts: HashMap<char, usize> = duplicate_character_counter(&s);
    show_duplicate_character_counts(&s, &counts);
}
